libnavigation-ui/src/main/java/com/mapbox/navigation/ui/speed/SpeedLimitView.java

```diff
@@ -0,0 +1,213 @@
+    if (shape == GradientDrawable.RECTANGLE) {
+      background.setCornerRadius(RADIUS);
+    }
+    return background;
+  }
+
+  private GradientDrawable borderDrawable(int shape) {
+    GradientDrawable border = new GradientDrawable();
+    border.setShape(shape);
+    if (shape == GradientDrawable.RECTANGLE) {
+      border.setCornerRadius(RADIUS);
+      border.setColor(MUTCD_BORDER_COLOR);
+    } else {
+      border.setColor(VIENNA_BORDER_COLOR);
+    }
+    return border;
+  }
+
+  private LayerDrawable getSpeedDrawable(SpeedLimitSign speedLimitSign) {
+    LayerDrawable layerDrawable;
+    if (speedLimitSign == SpeedLimitSign.MUTCD) {
+      GradientDrawable[] layers = {
+          backgroundDrawable(GradientDrawable.RECTANGLE),
+          borderDrawable(GradientDrawable.RECTANGLE),
+          backgroundDrawable(GradientDrawable.RECTANGLE)
+      };
+      layerDrawable = new LayerDrawable(layers);
+      layerDrawable.setLayerInset(
+          0,
+          VIENNA_MUTCD_OUTER_BACKGROUND_INSET,
+          VIENNA_MUTCD_OUTER_BACKGROUND_INSET,
+          VIENNA_MUTCD_OUTER_BACKGROUND_INSET,
+          VIENNA_MUTCD_OUTER_BACKGROUND_INSET
+      );
+      layerDrawable.setLayerInset(
+          1,
+          VIENNA_MUTCD_BORDER_INSET,
+          VIENNA_MUTCD_BORDER_INSET,
+          VIENNA_MUTCD_BORDER_INSET,
+          VIENNA_MUTCD_BORDER_INSET
+      );
+      layerDrawable.setLayerInset(
+          2,
+          MUTCD_INNER_BACKGROUND_INSET,
+          MUTCD_INNER_BACKGROUND_INSET,
+          MUTCD_INNER_BACKGROUND_INSET,
+          MUTCD_INNER_BACKGROUND_INSET
+      );
+    } else {
+      GradientDrawable[] layers = {
+          backgroundDrawable(GradientDrawable.OVAL),
+          borderDrawable(GradientDrawable.OVAL),
+          backgroundDrawable(GradientDrawable.OVAL)
+      };
+      layerDrawable = new LayerDrawable(layers);
+      layerDrawable.setLayerInset(
+          0,
+          VIENNA_MUTCD_OUTER_BACKGROUND_INSET,
+          VIENNA_MUTCD_OUTER_BACKGROUND_INSET,
+          VIENNA_MUTCD_OUTER_BACKGROUND_INSET,
+          VIENNA_MUTCD_OUTER_BACKGROUND_INSET
+      );
+      layerDrawable.setLayerInset(
+          1,
+          VIENNA_MUTCD_BORDER_INSET,
+          VIENNA_MUTCD_BORDER_INSET,
+          VIENNA_MUTCD_BORDER_INSET,
+          VIENNA_MUTCD_BORDER_INSET
+      );
+      layerDrawable.setLayerInset(
+          2,
+          VIENNA_INNER_BACKGROUND_INSET,
+          VIENNA_INNER_BACKGROUND_INSET,
+          VIENNA_INNER_BACKGROUND_INSET,
+          VIENNA_INNER_BACKGROUND_INSET
+      );
+    }
+    return layerDrawable;
+  }
+
+  @SuppressLint("DefaultLocale")
+  private String getSpeedLimit(SpeedLimitSign sign, SpeedLimitUnit unit, int speedLimitKmph) {
+    if (sign == SpeedLimitSign.MUTCD) {
+      if (unit == SpeedLimitUnit.KILOMETRES_PER_HOUR) {
+        return getContext().getString(R.string.max_speed, speedLimitKmph);
+      } else {
+        double speed = Math.round(speedLimitKmph * KILO_MILES_FACTOR);
```

Comment:
If I remember correctly, there’s a risk of the conversion being off by one if the speed limit started out specified in miles per hour in the original dataset, got converted to kilometers per hour for storage in the routing tiles, and got converted back to miles per hour for display. Since we can’t wait for the routing tiles to be extended with support for the speed limit as it exists in the real world, the iOS implementation is avoiding any potential off-by-one error by [rounding to the nearest multiple of five](https://github.com/mapbox/mapbox-navigation-ios/blob/ae18aaa66a1b752dbbe62d4402a857b7a3a96f60/MapboxNavigation/SpeedLimitView.swift#L68-L69).

This does prevent us from accurately depicting nonmultiples of five, such as 7½ or 7.5, but those nonstandard speed limits almost always occur on private service roads, such as in parking lots or construction areas, as a cheeky way of telling drivers to slow down. Being off by 2½ in those cases is not as critical as being off by one on public roads.

Reply:
Done 👍